### DISCUSSION TITLE
Using the commit hooks on Windows

### DISCUSSION THREAD

**Discussion by Josverl** [micropython/micropython#12045] [Core Development] [answered]
I have been struggling to get the commit hooks to work on Windows 

 * Python 3.11.4 
   installed from the Wwindows store 
   ![image](https://github.com/micropython/micropython/assets/981654/049b6c00-60ac-4a65-a3c5-6f5502c90921)
* pre-commit  3.3.3
   Installed from PyPi
* Installed the pre-commit hooks with `pre-commit install --hook-type pre-commit --hook-type commit-msg`  as per the docs 
* Installed uncrustify 0.72 

if I test the pre-commit hooks with:  `pre-commit run --files .\tools\mpremote\mpremote\main.py`

then the output is : 
```
MicroPython codeformat.py for changed files..............................Failed
- hook id: codeformat
- exit code: 1

Executable `python3` not found

ruff.....................................................................Passed
```

Clearly `python3` cannot be found , and that is to be expected as that is not a valid way to start Python on Windows.

Proper ways to specify a manor/minor version are : `python -V:3` or 'py -3` or many other ways, however there is no python3 or python3.exe 
ref: https://docs.python.org/3.11/using/windows.html 

Running codeformat.py directly works fine:

```
python tools/codeformat.py -v -f .\tools\mpremote\mpremote\main.py 
Identified `C:\develop\MyPython\micropython\tools\mpremote` as project root containing a pyproject.toml.
Sources to be formatted: "mpremote/main.py"
Using configuration from project root.
target_version: ['py34', 'py35', 'py36', 'py37', 'py38', 'py39', 'py310', 'py311']
C:\develop\MyPython\micropython\tools\mpremote\mpremote\main.py already well formatted, good job.
```

Shebang lines are also supported:  https://docs.python.org/3.11/using/windows.html#shebang-lines
However if I change the Shebang to '#!/usr/bin/env python' ( so dropping the 3)  suddenly the hooks start to work 

So I'm in the dark of how this is supposed to work ( and yes, it works well in *nix , WSL ,containers and in codespaces, so there are work-arounds aplenty)  
is this a problem in the hook definition , or is this something that should be raised against the [pre-commit repo](https://github.com/pre-commit/pre-commit)

So : Has anyone gotten this to work on Windows?

**Comment by stinos** [comment]
> python -V:3 

That should also be `py -V:3` I guess, because when you write `python` it's expected that is the actual binary (which has no such argument), not the launcher.

> However if I change the Shebang to '#!/usr/bin/env python' ( so dropping the 3) suddenly the hooks start to work

Don't have time to try this now, but: I think that would mean that pre-commit is using the launcher `py`, so making it use python directly instead might fix things. What's in your .git/hooks/pre-commit ? Maybe that one can be changed, or there's a way to tell `pre-commit install` what to use.

**Reply by Josverl** [reply, marked as answer]
did some rip and rebuild to figure this out:
- Uninstalled all python version and launchers
- Found that there was a python3 > py.exe powershell alias defined for powershell. ( get-alias python) 
   I must have created this in an earlier attempt to get this to work. Flash is not the only memory that can degrade apparently. (that was also why python _appeared_ to accept the -V:3 , that was actually py.exe'

Then : 
- Install python 3.11 from the Windows store 
   ```
   > python --version  
   Python 3.11.4
   > python3 --version 
    python3 : The term 'python3' is not recognized as the name of a cmdlet, function, script file, or operable program
   ```
- checked Windows Application Aliases ( in Settings) , and found that the 'App aliases' for python3 and python3.11 were not enabled.
   ![image](https://github.com/micropython/micropython/assets/981654/58c51dc4-5822-4d8b-922c-9070b33a2fd6)
 
- Enabled the alias for `python3` 
   > python3 --version  
   Python 3.11.4
   ```
- created a new venv 
- installed pre-commit and the hooks 
- found the the pre-commit hook now runs as expected
```
MicroPython codeformat.py for changed files..............................Passed
- hook id: codeformat
- duration: 1.02s

Identified `C:\develop\MyPython\micropython\tools\mpremote` as project root containing a pyproject.toml.
Using configuration from project root.
target_version: ['py34', 'py35', 'py36', 'py37', 'py38', 'py39', 'py310', 'py311', 'py312']
Found input source: "mpremote/main.py"
C:\develop\MyPython\micropython\tools\mpremote\mpremote\main.py already well formatted, good job.

All done! \u2728 \U0001f370 \u2728
1 file left unchanged.

ruff.....................................................................Passed
```
success!

Root cause : attempts to create a python3 alias in powershell .

Further testing also shows that the python3 Windows App Alias is indeed required.
![image](https://github.com/micropython/micropython/assets/981654/53f45861-0291-4484-9cf6-0029626870b7)

**Reply by scruss** [reply]
> ... the 'App aliases' for python3 and python3.11 were not enabled.

The Windows convention for naming Python 3 is seemingly at odds with the Linux and Mac standards: the command `python` shouldn't launch any interpreter on those systems, and `python3` is the one to use. If there's still the obsolete Python 2 interpreter on your system, that's the one launched by the `python` command.

**Reply by Josverl** [reply]
The convention was created by the folks at python.org. See links above. 
The python 3 app aliasses are curtesy of the store app, they are part of the application package. 

No python 2 needed, nor installed in the last 3 years.

But hey I always think the dashboard of my neighbor's car looks a bit strange as well 😉

**Reply by stinos** [reply]
FWIW I tried this in an anaconda environment on windows and that has similar issues. Even though pre-commit knows which python version I'd want it to use (it's full path is in the hooks\pre-commit file after installing), when running it wants a python3 executable which I do not have. I don't immediately see a way to make it use python instead, but a hacky workaround is to symlink python3 to python.

**Reply by Josverl** [reply]
> symlink python3 to python.

I think that is what the windows store aliasses do under the covers

**Comment by Josverl** [comment]
After re-install python yet again - but now not form the store - i ran into this issue again 


Documenting the steps to to create the `python3.exe` Symlink 

Run in a Elevated powershell prompt.
```pwsh
# adjust the path for your version of python
cd $env:userprofile\AppData\Local\Programs\Python\Python311 
new-item python3.exe -itemtype SymbolicLink -Value python.exe
```
